### PR TITLE
Extract openai API calls and retry at lowest level

### PR DESCRIPTION
**Pull request by collijk**
### Background
Fourth PR in the chain to organize the LLM interaction.  See:
- #3191 
- #3436 
- #3486 

In a previous PR, I extracted a retry decorator from the embedding calls.  This PR uses the same retry decorator on an extracted call to create chat completions from the OpenAI API.  

### Changes
- APIManager now only manages API budgets.  Will be renamed and updated to use LLM structs in the next PR.
- `autogpt.llm.llm_utils.retry_openai_api` moved to `autogpt.llm.providers.openai.retry_api`
- Low level API methods for creating embeddings and chat completions have been extracted to `autogpt.llm.providers.openai` and have been appropriately decorated to consistently handle API failure cases.
- `autogpt.llm.llm_utils.create_chat_completion` refactored to use low-level wrapped api calls and update the budget manually.
- Move retry handler tests from `tests.unit.test_llm_utils` to `tests.unit.test_openai`

### Documentation
N/A, this is primarily code motion

### Test Plan
N/A this is primarily code motion and the code in question has unit tests.

### PR Quality Checklist
- [X] My pull request is atomic and focuses on a single change.
- [X] I have thoroughly tested my changes with multiple different prompts.
- [X] I have considered potential risks and mitigations for my changes.
- [X] I have documented my changes clearly and comprehensively.
- [X] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Review by Sudo-psc — DISMISSED** [review]

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Comment by lc0rp** [comment]
James, @collijk - I know you've got your head down. Is this worth trying to put in 0.3.2? Is anyone else available to resolve conflicts?

**Comment by lc0rp** [comment]
Seems to me, this could still be valuable even after the re-arch. Marking for maintainer-review

**Comment by ntindle** [comment]
Needs rebasing

**Comment by p-i-** [comment]
What's going on with:

> Some checks were not successful
> Docker CI / test (pull_request) Cancelled after 327m
> Python CI / test (3.10) (pull_request_target) Cancelled after 329m

Are we racking up costs here?

Should we have a safeguard to abort after say 300s?

File: autogpt/llm/utils/__init__.py [hunk 1]
```diff
@@ -209,41 +111,48 @@ def create_chat_completion(
...
-        max_tokens=max_tokens,
+        **chat_completion_kwargs,
     )
+    logger.debug(f"Response: {response}")
+
+    resp = ""
+    if not hasattr(response, "error"):
+        resp = response.choices[0].message["content"]
+    else:
+        resp = str(response.error)
```
Comment (Pwuts) [inline autogpt/llm/utils/__init__.py]: The return value can either be the text from a chat completion or an error message (not sure what the error looks like). Probably better to raise an error if one occurs?

File: tests/unit/test_api_manager.py [hunk 2]
```diff
@@ -27,67 +27,69 @@ def mock_costs():
...
-            mock_response.usage.completion_tokens = 20
-            mock_create.return_value = mock_response
-
-            api_manager.create_chat_completion(messages, model=model)
-
-            assert api_manager.get_total_prompt_tokens() == 10
-            assert api_manager.get_total_completion_tokens() == 20
-            assert api_manager.get_total_cost() == (10 * 0.002 + 20 * 0.002) / 1000
+    # TODO: Port these into a combination of integration tests for the API and
+    #  unit tests for the ApiManager class.
```
Comment (Pwuts) [inline tests/unit/test_api_manager.py]: Is this a TODO before merging this PR?

File: tests/unit/test_openai.py
Comment (Pwuts) [inline tests/unit/test_openai.py:1]: test_retry_{`open_api`, `openapi` or `openai`?}

File: tests/unit/test_openai.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    assert result == call_count
+    assert raises.count == call_count
+
+    output = capsys.readouterr()
+
+    assert "Reached rate limit, passing..." in output.out
+    assert "Please double check" not in output.out
+
+
+def test_retry_openapi_other_api_error(capsys):
```
Comment (Pwuts) [inline tests/unit/test_openai.py]: nit: type hint for `capsys`

File: tests/unit/test_openai.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        if type(error) == RateLimitError:
+            assert "Reached rate limit, passing..." in output.out
+            assert "Please double check" in output.out
+        if type(error) == APIError:
+            assert "API Bad gateway" in output.out
+    else:
+        assert output.out == ""
+
+
+def test_retry_open_api_rate_limit_no_warn(capsys):
```
Comment (Pwuts) [inline tests/unit/test_openai.py]: nit: parameter type hint

File: tests/unit/test_openai.py
Comment (Pwuts) [inline tests/unit/test_openai.py:1]: It isn't obvious to me what each of these tests does and tests for, with the exception of `#_no_error`. Could you clarify the test names or add docstrings?

File: autogpt/llm/utils/__init__.py (hunk 1 above)
Comment (ntindle) [reply autogpt/llm/utils/__init__.py]: Thats fair, will swap it for a raise

File: tests/unit/test_api_manager.py (hunk 2 above)
Comment (ntindle) [reply tests/unit/test_api_manager.py]: Really up to @collijk but I was going to work on it tomorrow

File: tests/unit/test_openai.py
Comment (ntindle) [reply tests/unit/test_openai.py:1]: test_retry_provider_openai?

File: autogpt/llm/providers/openai.py [hunk 5]
```diff
@@ -72,3 +91,157 @@
...
+            logger.warn(f"Failed to update API costs: {err.__class__.__name__}: {err}")
+
+    def metering_wrapper(*args, **kwargs):
+        openai_obj = openai_obj_processor(*args, **kwargs)
+        if isinstance(openai_obj, OpenAIObject) and "usage" in openai_obj:
+            update_usage_with_response(openai_obj)
+        return openai_obj
+
+    def metered_func(*args, **kwargs):
+        with patch.object(
```
Comment (erik-megarad) [inline autogpt/llm/providers/openai.py:121]: Feels unusual to be using `unittest.mock.patch` in non-test code

File: autogpt/llm/utils/__init__.py [hunk 6]
```diff
@@ -209,41 +111,48 @@ def create_chat_completion(
     logger.debug(
         f"{Fore.GREEN}Creating chat completion with model {model}, temperature {temperature}, max_tokens {max_tokens}{Fore.RESET}"
     )
+    chat_completion_kwargs = {
```
Comment (erik-megarad) [inline autogpt/llm/utils/__init__.py:114]: Love it

**Review by erik-megarad — COMMENTED** [review]
Only did a high-level readthrough but I like these changes

File: tests/unit/test_api_manager.py (hunk 2 above)
Comment (ntindle) [reply tests/unit/test_api_manager.py]: Done, mostly

File: tests/integration/test_provider_openai.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            del mock_response.error
+            mock_response.usage.prompt_tokens = 10
+            mock_response.usage.completion_tokens = 20
+            mock_create.return_value = mock_response
+
+            openai.create_chat_completion(messages, model=model)
+
+            assert api_manager.get_total_prompt_tokens() == 10
+            assert api_manager.get_total_completion_tokens() == 20
+            assert api_manager.get_total_cost() == (10 * 0.002 + 20 * 0.002) / 1000
```
Comment (ntindle) [inline tests/integration/test_provider_openai.py]: @collijk Can you point me in the right direction on this one?

I'm not sure how to actually evaluate this is called without standing up the whole system

Comment (collijk) [reply tests/integration/test_provider_openai.py]: I'd just ditch it then.  It's not a super important test.  And it's no longer a unit test.

File: autogpt/llm/providers/openai.py (hunk 5 above)
Comment (Pwuts) [reply autogpt/llm/providers/openai.py:121]: This code is just being moved here from `llm.utils`. But you're right, it is unusual :)

**Comment by Pwuts** [comment]
@ntindle status?

**Comment by ntindle** [comment]
There's a test error I've been banging my head against. I think it was related to my setup though

**Review by merwanehamadi — APPROVED** [review]

**Pull request merged**